### PR TITLE
Recursively transform predicates in `ty::Dynamic` rather than using identity

### PR DESCRIPTION
**Pull request by compiler-errors**
It is not correct to be using the *identity* trait ref for a `dyn` type. This leads to us downstream not being able to assert the non-existence of things like `ty::Alias`, `ty::Param`, `ty::ReEarlyParam`, etc. and also doesn't allow us to distinguish, e.g., `dyn Trait<i32>` and `dyn Trait<&str>`, lol.

Instead, just recursively transform the `dyn` predicates. We also should definitely be keeping around the projection predicates too.

cc @rcvalle @maurer

### CODE REVIEW

**Comment by compiler-errors** [comment]
Is there a way to bulk fix the `tests/codegen/sanitizer/cfi/*.rs` tests? They're kind of excruciating to edit.

**Comment by rcvalle** [comment]
I agree with this approach. I just didn't know it was possible/how to do it at the time I implemented it using the identity of the trait that implemented the method.

**Comment by maurer** [comment]
This looks reasonable to me as well.

**Comment by compiler-errors** [comment]
Ok well it's gonna take approximately 5 millennia to bless those tests then 💀

**Comment by rcvalle** [comment]
No worries! I can work on that and send a PR later.

**Comment by maurer** [comment]
This raised a thought in my mind - once we're done getting everything working, should `transform_ty` maybe be rewritten to use a type visitor? It seems like we would have gotten this one right "for free" without needing compiler_errors' wisdom if we had been using a visitor instead of manual recursion.

**Comment by compiler-errors** [comment]
Yes, it could be implemented as a `TypeFolder`, though we'd need to be careful to walk types correctly given those transform options. Still very doable.

**Comment by rcvalle** [comment]
I don't know, maybe? If we can do depth-first transformation of types using it, I guess it should work. If it's an improvement, I'm definitively up to it later.

**Comment by rcvalle** [comment]
FWIW, I don't know how type visitor works in detail, but how we do it now at least forces us to explicitly maintain the types we support encoding for, and are let known otherwise.

**Comment by compiler-errors** [comment]
I strongly agree that this should be rewritten as a `TypeFolder`. If you want to guarantee you're actually transforming the types correctly for each specific variant, then just use an exhaustive match. That is off-topic for this PR, though.

File: compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs [hunk 1]
```diff
@@ -746,22 +746,40 @@ fn encode_ty<'tcx>(
...
+                    term: match term.unpack() {
+                        TermKind::Ty(ty) => transform_ty(tcx, ty, parents, options).into(),
+                        TermKind::Const(ct) => ct.into(),
+                    },
+                }),
+                ty::ExistentialPredicate::AutoTrait(..) => predicate,
+            })
+        })
+        .collect();
+    predicates.sort_by(|a, b| a.skip_binder().stable_cmp(tcx, &b.skip_binder()));
```
Comment (lcnr) [inline compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs:781]: why is this needed? please add a comment why we have to sort here

**Review by lcnr — COMMENTED** [review]
lacking some context, will review once I am back from RustNation

File: compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs:781]: Actually, this doesn't matter. We only sort by def path rn lol

**Comment by compiler-errors** [comment]
this pr is stuck in review hell and @maurer has volunteered to just do it themselves and i'm just gonna r+ that
